src/Spice86.Core/Emulator/InterruptHandlers/Dos/Ems/ExpandedMemoryManager.cs

```diff
@@ -0,0 +1,1220 @@
+﻿namespace Spice86.Core.Emulator.InterruptHandlers.Dos.Ems;
+
+using Serilog.Events;
+
+using Spice86.Core.Emulator.Callback;
+using Spice86.Core.Emulator.Errors;
+using Spice86.Core.Emulator.InterruptHandlers;
+using Spice86.Core.Emulator.Memory;
+using Spice86.Core.Emulator.OperatingSystem.Devices;
+using Spice86.Core.Emulator.OperatingSystem.Enums;
+using Spice86.Core.Emulator.VM;
+using Spice86.Shared.Emulator.Errors;
+using Spice86.Shared.Interfaces;
+using Spice86.Shared.Utils;
+
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+/// <summary>
+/// Provides DOS applications with EMS memory.
+/// </summary>
+public sealed class ExpandedMemoryManager : InterruptHandler, IMemoryDevice {
+    public const string EmsIdentifier = "EMMXXXX0";
+
+    /// <summary>
+    /// EMM system handle (reserved for OS usage)
+    /// </summary>
+    public const byte EmmSystemHandle = 0;
+
+    public const byte EmmMaxHandles = 16;
+
+    public const uint EmmPageFrameSize = 64 * 1024;
+
+    public const byte EmmMaxPhysicalPages = (byte) (EmmPageFrameSize / EmmPageSize);
+
+    public const ushort EmmNullPage = 0xFFFF;
+    
+    public const ushort EmmNullHandle = 0xFFFF;
+
+    public const ushort EmmPageFrameSegment = 0xE000;
+
+    public const ushort EmmPageSize = 16384;
+
+    public override ushort? InterruptHandlerSegment => 0xF100;
+    
+    public override byte Index => 0x67;
+
+    private readonly ILoggerService _loggerService;
+    
+    public MemoryBlock MemoryBlock { get; }
+
+    public EmmMapping[] EmmSegmentMappings { get; } = new EmmMapping[64];
+
+    public EmmMapping[] EmmMappings { get; } = new EmmMapping[EmmHandle.EmmMaxPhysicalPages];
+    
+    public EmmHandle[] EmmHandles { get; } = new EmmHandle[EmmMaxHandles];
+    
+    public ushort MemorySizeInMb { get; }
+
+    public int TotalPages => MemoryBlock.Pages;
+
+    private readonly Ram _ram;
+
+    public ushort GetFreePages() => Math.Min((ushort) 0x7FFF, (ushort) (GetFreeMemoryTotal() / 4));
+
+    public ExpandedMemoryManager(Machine machine, ILoggerService loggerService) : base(machine, loggerService) {
+        _loggerService = loggerService;
+        MemorySizeInMb = Math.Max((ushort)8, (ushort) (_memory.Size / 1024 / 1024));
+        _ram = new Ram((uint) (MemorySizeInMb * 1024 * 1024));
+        var device = new CharacterDevice(DeviceAttributes.Ioctl, EmsIdentifier);
+        machine.Dos.AddDevice(device, InterruptHandlerSegment, 0x0000);
```

Comment:
Not sure why you need to put that in the bios segment (F100 is ROM on the IBM PC)

Reply:
Removed.